Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,324 @@
+//
+//  AuthenticationInterceptor.swift
+//
+//  Copyright (c) 2020 Alamofire Software Foundation (http://alamofire.org/)
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+/// NOTE: CN docstring
+public protocol AuthenticationCredential {
+    /// NOTE: CN docstring
+    var requiresRefresh: Bool { get }
+}
+
+// MARK: -
+
+/// NOTE: CN docstring
+public protocol Authenticator: AnyObject {
+    /// NOTE: CN docstring
+    associatedtype AuthCredential: AuthenticationCredential
+
+    /// NOTE: CN rework docstring
+    ///
+    /// Authenticates the `URLRequest` with the `Credential` information. In the case of OAuth2, the access token of
+    /// the `Credential` would be added as a Bearer token to the `Authorization` header.
+    func apply(_ credential: AuthCredential, to urlRequest: inout URLRequest)
+
+    /// NOTE: CN rework docstring
+    ///
+    /// Refreshes the `AuthCredential` and executes the completion closure with the `Result` once complete.
+    func refresh(_ credential: AuthCredential, completion: @escaping (Result<AuthCredential, Error>) -> Void)
```

Comment:
Could it be useful to also hand over the `Session` here? That way an `Authenticator` could use this session the perform a refresh.

Reply:
Good point. No real reason not to pass it through. Addressed in d7b4d1d1.